### PR TITLE
[GOBBLIN-260] Fix Salesforce dynamic partitioning bugs

### PR DESCRIPTION
**Pull request by jack-moseley**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-260

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

1. Use low watermark in histogram query for incremental job with dynamic partitioning
2. Don't do dynamic partitioning if source.max.number.of.partitions = 1

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by jack-moseley** [comment]
@htran1 @zxcware please review

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 1]
```diff
@@ -80,6 +80,8 @@
   private static final String ENABLE_DYNAMIC_PARTITIONING = "salesforce.enableDynamicPartitioning";
   private static final String DAY_PARTITION_QUERY_TEMPLATE = "SELECT count(${column}) cnt, DAY_ONLY(${column}) time FROM ${table} "
       + "WHERE CALENDAR_YEAR(${column}) = ${year} GROUP BY DAY_ONLY(${column}) ORDER BY DAY_ONLY(${column})";
+  private static final String DAY_PARTITION_QUERY_LOW_WATERMARK_TEMPLATE = "SELECT count(${column}) cnt, DAY_ONLY(${column}) time FROM ${table} "
```
[the remark this answers is not in the document]
Comment (jack-moseley) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Using high watermark as upper bound for the last year now

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 2]
```diff
@@ -80,6 +80,8 @@
   private static final String ENABLE_DYNAMIC_PARTITIONING = "salesforce.enableDynamicPartitioning";
   private static final String DAY_PARTITION_QUERY_TEMPLATE = "SELECT count(${column}) cnt, DAY_ONLY(${column}) time FROM ${table} "
```
[the remark this answers is not in the document]
Comment (jack-moseley) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Was using it for years outside of the low watermark year, but now changed it to only need query

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 3]
```diff
@@ -209,7 +212,27 @@ private Histogram getHistogram(String entity, String watermarkColumn, SourceStat
     }
 
     for (int year = startYear; year <= endYear; year++) {
-      values.put("year", Integer.toString(year));
+
+      if (year == startYear) {
+        values.put("start", lowWatermarkDate);
+        values.put("greater", partition.isLowWatermarkInclusive() ? ">=" : ">");
+      } else {
+        calendar.clear();
```
Comment (zxcware) [inline gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: We can define a help function which returns the time string given a year integer

Comment (jack-moseley) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Created getDateString helper

**Pull request closed without merging**